File: 10. Stack and Queue/4. Implementation Problems/Sliding-Window-Maximum.py
# Sliding Window Maximum (Leetcode)
# Leetcode Link : https://leetcode.com/problems/sliding-window-maximum/description/

# Time Complexity : O(n)
# Space Complexity : O(n)

from collections import deque
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        n = len(nums)
        que = deque() # Deque to store indices of the elements
        ans = []

        for i in range(n):
            # While que is not empty and element at indices at last of the que is >= to the current element
            while que and nums[que[-1]] <= nums[i]:
                # Pop the index from the rear
                que.pop()
            
            # Append the current index in the que
            que.append(i)

            # Remove the first element of que if it is outside the window
            if que[0] == i - k:
                # Pop the index front the front
                que.popleft()
            
            # If the window has k elements, append the max element of the current window in the ans
            if i >= k - 1:
                ans.append(nums[que[0]])
        
        return ans

File: 16. Sliding Window & Two Pointers/2. Medium Problems/Sliding-Window-Maximum.cpp
// Sliding Window Maximum (Leetcode)
// Leetcode Link : https://leetcode.com/problems/sliding-window-maximum/description/

// Time Complexity : O(n)
// Space Complexity : O(n)

class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        vector<int> maximums;
        list<int> lst;

        int left = 0;
        for(int right = 0; right < nums.size(); right++){
            // If the window exceeds k, shrink the window
            if(k < (right-left+1)){
                // If the element leaving the window is the same as the maximum, remove it from the list
                if(nums[left] == lst.front()) lst.pop_front();
                left++; // Move the left boundary to the right
            }
            
            // While list is not empty and elements from the back of the list are <= the current element
            while(!lst.empty() && nums[right] > lst.back()){
                lst.pop_back(); // Remove it from the list
            }
            // Add the current element to the back of the list
            lst.push_back(nums[right]);

            // If the window size has reached k
            if(k == (right-left+1)){
                // The maximum of the current window is the front of the list
                maximums.push_back(lst.front());
            }
        }

        return maximums;
    }
};